src/main/java/reposense/authorship/analyzer/AuthorshipAnalyzer.java

```diff
@@ -0,0 +1,269 @@
+            int currentPreImageLineNumber = getPreImageStartingLineNumber(linesChanged[LINES_CHANGED_HEADER_INDEX]);
+
+            // skip the lines changed header, index starts from 1
+            for (int lineIndex = 1; lineIndex < linesChanged.length; lineIndex++) {
+                String lineChanged = linesChanged[lineIndex];
+
+                if (lineChanged.startsWith(DELETED_LINE_SYMBOL)) {
+                    String deletedLineContent = lineChanged.substring(DELETED_LINE_SYMBOL.length());
+                    double lowestOriginalityScore = lowestOriginalityLine == null
+                            ? Integer.MAX_VALUE
+                            : lowestOriginalityLine.getOriginalityScore();
+                    double originalityScore = computeOriginalityScore(lineContent, deletedLineContent,
+                            lowestOriginalityScore);
+
+                    if (lowestOriginalityLine == null
+                            || originalityScore < lowestOriginalityLine.getOriginalityScore()) {
+                        lowestOriginalityLine = new CandidateLine(
+                                currentPreImageLineNumber, deletedLineContent, filePath, commitHash,
+                                originalityScore);
+                    }
+                }
+
+                if (!lineChanged.startsWith(ADDED_LINE_SYMBOL)) {
+                    currentPreImageLineNumber++;
+                }
+            }
+        }
+
+        return lowestOriginalityLine;
+    }
+
+    /**
+     * Returns the pre-image starting line number by matching the pattern inside {@code linesChangedHeader}.
+     *
+     * @throws AssertionError if lines changed header matcher failed to find anything.
+     */
+    private static int getPreImageStartingLineNumber(String linesChangedHeader) {
+        Matcher linesChangedHeaderMatcher = STARTING_LINE_NUMBER_PATTERN.matcher(linesChangedHeader);
+
+        if (!linesChangedHeaderMatcher.find()) {
+            logger.severe(
+                    String.format(MATCH_GROUP_FAIL_MESSAGE_FORMAT, PREIMAGE_START_LINE_GROUP_NAME, linesChangedHeader));
+            throw new AssertionError(
+                    "Should not have error matching line number pattern inside lines changed header!");
+        }
+
+        return Integer.parseInt(linesChangedHeaderMatcher.group(PREIMAGE_START_LINE_GROUP_NAME));
+    }
+
+    /**
+     * Calculates the originality score of {@code s} with {@code baseString}.
+     */
+    private static double computeOriginalityScore(String s, String baseString, double limit) {
+        double levenshteinDistance = StringsUtil.getLevenshteinDistance(s, baseString, limit * baseString.length());
+        return levenshteinDistance / baseString.length();
+    }
+
+    /**
+     * Returns the git blame line info for {@code line}.
+     */
+    private static GitBlameLineInfo getGitBlameLineInfo(RepoConfiguration config, CandidateLine line) {
```

Comment:
I can move it in but the update for `aggregateBlameAuthorModifiedAndDateInfo` would be a separate PR as mentioned above.

Reply:
I personally prefer that `GitBlameLineInfo` stores author information as something like a (name, email) tuple, so that the user isn't obliged to have a way to generate an author in order to use it. We could perhaps then overload `Author getAuthor(BiFunction<String, String, Author> authorGenerator)` in `GitBlameLineInfo` as you've described.

Reply:
I think let's just store the raw author info, then call `config.getAuthor()` when needed.